plugins/uefi-capsule/fu-uefi-cod-device.c

```diff
@@ -0,0 +1,190 @@
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-uefi-cod-device.h"
+#include "fu-uefi-common.h"
+
+struct _FuUefiCodDevice
+{
+	FuUefiDevice		 parent_instance;
+};
+
+G_DEFINE_TYPE (FuUefiCodDevice, fu_uefi_cod_device, FU_TYPE_UEFI_DEVICE)
+
+static gboolean
+fu_uefi_cod_device_get_results_for_idx (FuDevice *device, guint idx, GError **error)
+{
+	FuUefiDevice *device_uefi = FU_UEFI_DEVICE (device);
+	const gchar *tmp;
+	fwupd_guid_t guid = { 0x0 };
+	gsize bufsz = 0;
+	guint32 status = 0;
+	guint32 total_size = 0;
+	g_autofree gchar *err_msg = NULL;
+	g_autofree gchar *guidstr = NULL;
+	g_autofree gchar *name = NULL;
+	g_autofree gchar *version_str = NULL;
+	g_autofree guint8 *buf = NULL;
+
+	/* lets assume we just did one update */
+	name = g_strdup_printf ("Capsule%04u", idx);
+	if (!fu_efivar_get_data (FU_EFIVAR_GUID_EFI_GLOBAL, name,
+				 &buf, &bufsz, NULL, error)) {
+		g_prefix_error (error, "failed to read EFI variable: ");
+		return FALSE;
+	}
+
+	/* sanity check */
+	if (!fu_common_read_uint32_safe (buf, bufsz, 0x00, &total_size, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (total_size < 0x3A) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "EFI_CAPSULE_RESULT_VARIABLE_HEADER too small");
+		return FALSE;
+	}
+
+	/* verify guid */
+	if (!fu_memcpy_safe (guid, sizeof(guid), 0x0,		/* dst */
+			     buf, bufsz, 0x08, sizeof(guid), error))
+		return FALSE;
+	guidstr = fwupd_guid_to_string (&guid, FWUPD_GUID_FLAG_MIXED_ENDIAN);
+	if (g_strcmp0 (guidstr, fu_uefi_device_get_guid (device_uefi)) != 0) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_FOUND,
+			     "wrong GUID, expected %s, got %s",
+			     fu_uefi_device_get_guid (device_uefi),
+			     guidstr);
+		return FALSE;
+	}
+
+	/* get status */
+	if (!fu_common_read_uint32_safe (buf, bufsz, 0x36, &status, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (status == FU_UEFI_DEVICE_STATUS_SUCCESS) {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_SUCCESS);
+		return TRUE;
+	}
+
+	/* something went wrong */
+	if (status == FU_UEFI_DEVICE_STATUS_ERROR_PWR_EVT_AC ||
+	    status == FU_UEFI_DEVICE_STATUS_ERROR_PWR_EVT_BATT) {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_FAILED_TRANSIENT);
+	} else {
+		fu_device_set_update_state (device, FWUPD_UPDATE_STATE_FAILED);
+	}
+	version_str = g_strdup_printf ("%u", fu_uefi_device_get_version_error (device_uefi));
+	tmp = fu_uefi_device_status_to_string (status);
+	if (tmp == NULL) {
+		err_msg = g_strdup_printf ("failed to update to %s",
+					   version_str);
+	} else {
+		err_msg = g_strdup_printf ("failed to update to %s: %s",
+					   version_str, tmp);
+	}
+	fu_device_set_update_error (device, err_msg);
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_cod_device_get_results (FuDevice *device, GError **error)
+{
+	/* this should be enough for any system */
+	for (guint i = 0; i < 1024; i++) {
+		g_autoptr(GError) error_local = NULL;
```

Comment:
The CapsuleLast will have the last capsule we updated, so instead of running the loop from 0 here, we must read CapsuleLast (which will have somelike like Capsule0002, and start from 2